File: shared/events.go
// Package events defines the shared event schema used across all Go services.
// Schema version: 1.0.0
package events

import "time"

// EventType constants — must match shared/events.json
const (
	EventAgentCreated   = "agent.created"
	EventAgentStarted   = "agent.started"
	EventAgentStopped   = "agent.stopped"
	EventAgentFailed    = "agent.failed"
	EventAgentCompleted = "agent.completed"
	EventAgentDeleted   = "agent.deleted"
	EventAgentHeartbeat = "agent.heartbeat"
)

// NATS subject constants
const (
	SubjectAgentsEvents  = "agents.events"
	SubjectLogsStream    = "logs.stream"
	SubjectEventsStream  = "events.stream"
	SubjectMetricsStream = "metrics.stream"
)

// BaseEvent is the canonical agent lifecycle event envelope.
type BaseEvent struct {
	EventType   string                 `json:"event_type"`
	AgentID     string                 `json:"agent_id"`
	ExecutionID string                 `json:"execution_id,omitempty"`
	Timestamp   string                 `json:"timestamp"`
	Source      string                 `json:"source"`
	Payload     map[string]interface{} `json:"payload,omitempty"`
}

// NewBaseEvent constructs a BaseEvent with current UTC timestamp.
func NewBaseEvent(eventType, agentID, source string) BaseEvent {
	return BaseEvent{
		EventType: eventType,
		AgentID:   agentID,
		Source:    source,
		Timestamp: time.Now().UTC().Format(time.RFC3339),
		Payload:   make(map[string]interface{}),
	}
}

// LogEvent is the canonical log entry schema.
type LogEvent struct {
	AgentID     string                 `json:"agent_id"`
	ExecutionID string                 `json:"execution_id,omitempty"`
	Message     string                 `json:"message"`
	Level       string                 `json:"level"`
	Source      string                 `json:"source,omitempty"`
	Timestamp   string                 `json:"timestamp"`
	Metadata    map[string]interface{} `json:"metadata,omitempty"`
}

// NewLogEvent creates a LogEvent at the current UTC time.
func NewLogEvent(agentID, message, level, source string) LogEvent {
	return LogEvent{
		AgentID:   agentID,
		Message:   message,
		Level:     level,
		Source:    source,
		Timestamp: time.Now().UTC().Format(time.RFC3339),
	}
}

// MetricEvent is the canonical metrics schema.
type MetricEvent struct {
	AgentID     string                 `json:"agent_id"`
	MetricName  string                 `json:"metric_name"`
	MetricValue float64                `json:"metric_value"`
	Labels      map[string]interface{} `json:"labels,omitempty"`
	Timestamp   string                 `json:"timestamp"`
}

// NewMetricEvent creates a MetricEvent at the current UTC time.
func NewMetricEvent(agentID, name string, value float64) MetricEvent {
	return MetricEvent{
		AgentID:     agentID,
		MetricName:  name,
		MetricValue: value,
		Timestamp:   time.Now().UTC().Format(time.RFC3339),
	}
}

// AgentStartedPayload holds container info for agent.started events.
type AgentStartedPayload struct {
	ContainerID   string `json:"container_id"`
	ContainerName string `json:"container_name"`
	Image         string `json:"image"`
}

// AgentStoppedPayload holds exit info for agent.stopped events.
type AgentStoppedPayload struct {
	ContainerID string `json:"container_id"`
	ExitCode    *int   `json:"exit_code,omitempty"`
}

// AgentFailedPayload holds error info for agent.failed events.
type AgentFailedPayload struct {
	Error       string `json:"error"`
	ContainerID string `json:"container_id,omitempty"`
	ExitCode    *int   `json:"exit_code,omitempty"`
}

// HeartbeatPayload holds runtime stats for agent.heartbeat events.
type HeartbeatPayload struct {
	CPUPercent    float64 `json:"cpu_percent"`
	MemoryMB      float64 `json:"memory_mb"`
	MemoryPercent float64 `json:"memory_percent"`
	UptimeSeconds float64 `json:"uptime_seconds"`
}

File: shared/events.py
"""Shared event schema definitions — used by FastAPI backend and agent runtime."""
from __future__ import annotations
from enum import Enum
from typing import Any, Optional
from pydantic import BaseModel, Field
from datetime import datetime, timezone
import uuid


class EventType(str, Enum):
    AGENT_CREATED   = "agent.created"
    AGENT_STARTED   = "agent.started"
    AGENT_STOPPED   = "agent.stopped"
    AGENT_FAILED    = "agent.failed"
    AGENT_COMPLETED = "agent.completed"
    AGENT_DELETED   = "agent.deleted"
    AGENT_HEARTBEAT = "agent.heartbeat"


class LogLevel(str, Enum):
    DEBUG   = "debug"
    INFO    = "info"
    WARN    = "warn"
    WARNING = "warning"
    ERROR   = "error"
    FATAL   = "fatal"


class NATSSubject(str, Enum):
    AGENTS_EVENTS  = "agents.events"
    LOGS_STREAM    = "logs.stream"
    EVENTS_STREAM  = "events.stream"
    METRICS_STREAM = "metrics.stream"


class BaseEvent(BaseModel):
    event_type:   str
    agent_id:     str
    execution_id: Optional[str] = None
    timestamp:    str = Field(default_factory=lambda: datetime.now(timezone.utc).isoformat())
    source:       str = "unknown"
    payload:      dict[str, Any] = Field(default_factory=dict)

    @classmethod
    def create(cls, event_type: EventType | str, agent_id: str, source: str, **payload: Any) -> "BaseEvent":
        return cls(
            event_type=str(event_type),
            agent_id=agent_id,
            source=source,
            payload=payload,
        )


class LogEvent(BaseModel):
    agent_id:     str
    execution_id: Optional[str] = None
    message:      str
    level:        str = "info"
    source:       Optional[str] = None
    timestamp:    str = Field(default_factory=lambda: datetime.now(timezone.utc).isoformat())
    metadata:     dict[str, Any] = Field(default_factory=dict)


class MetricEvent(BaseModel):
    agent_id:     str
    metric_name:  str
    metric_value: float
    labels:       dict[str, Any] = Field(default_factory=dict)
    timestamp:    str = Field(default_factory=lambda: datetime.now(timezone.utc).isoformat())


class AgentStartedPayload(BaseModel):
    container_id:   str
    container_name: str
    image:          str


class AgentStoppedPayload(BaseModel):
    container_id: str
    exit_code:    Optional[int] = None


class AgentFailedPayload(BaseModel):
    error:        str
    container_id: Optional[str] = None
    exit_code:    Optional[int] = None


class AgentHeartbeatPayload(BaseModel):
    cpu_percent:    float
    memory_mb:      float
    memory_percent: float
    uptime_seconds: float